src/include/primer/p0_starter.h

```diff
@@ -0,0 +1,119 @@
+//===----------------------------------------------------------------------===//
+//
+//                         BusTub
+//
+// p0_starter.h
+//
+// Identification: src/include/primer/p0_starter.h
+//
+// Copyright (c) 2015-2020, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <memory>
+
+/*
+ * The base class defining a Matrix
+ */
+template <typename T>
+class Matrix {
+ protected:
+  // P0: Add implementation
+  Matrix(int r, int c) {}
+
+  // # of rows in the matrix
+  int rows;
+  // # of Columns in the matrix
+  int cols;
+  // Flattened array containing the elements of the matrix
+  // P0: Allocate the array in the constructor. Don't forget to free up
+  // the array in the destructor.
+  T *linear;
+
+ public:
+  // Return the # of rows in the matrix
+  virtual int getRows() = 0;
+
+  // Return the # of columns in the matrix
+  virtual int getColumns() = 0;
+
+  // Return the (i,j)th  matrix element
+  virtual T getElem(int i, int j) = 0;
+
+  // Sets the (i,j)th  matrix element to val
+  virtual void setElem(int i, int j, T val) = 0;
+
+  // Sets the matrix elements based on the array arr
+  virtual void matImport(T *arr) = 0;
+
+  // P0: Add implementation
+  virtual ~Matrix() = default;
+};
+
+template <typename T>
+class RowMatrix : public Matrix<T> {
+ public:
+  // P0: Add implementation
+  RowMatrix(int r, int c) : Matrix<T>(r, c) {}
+
+  // P0: Add implementation
+  int getRows() override { return 0; }
+
+  // P0: Add implementation
+  int getColumns() override { return 0; }
+
+  // P0: Add implementation
+  T getElem(int i, int j) override { return data[i][j]; }
+
+  // P0: Add implementation
+  void setElem(int i, int j, T val) override {}
+
+  // P0: Add implementation
+  void matImport(T *arr) override {}
+
+  // P0: Add implementation
+  ~RowMatrix() override = default;
+
+ private:
+  // 2D array containing the elements of the matrix in row-major format
+  // P0: Allocate the array of row pointers in the constructor. Use these pointers
+  // to point to corresponding elements of the 'linear' array.
+  // data should contain pointers to the rows in the linear array to make elements addressable using two dimensions.
+  // NOTE : Do not duplicate elements stored in linear.
+  // Don't forget to free up the array in the destructor.
+  T **data;
```

Comment:
Nit: I'm not sure if we have the same clang-tidy settings as terrier, but generally private members are postfixed with an underscore, `data_`.